### PR TITLE
task/refactor-state-machine-to-create-pause-state

### PR DESCRIPTION
**Pull request by tsaubergine**
Goal: Reduce number of ReacquireGPS and IMURestart states to improve extensibility of state machine, readability of state machine diagram, and simplify logic. Add ability for operator to "pause" mission and resume from where it was left off (like STOP, but not permanent).

Implements https://jaia-innovation.atlassian.net/browse/JAIAB-743?atlOrigin=eyJpIjoiNzBhNDZmNWE5YTRhNDVlM2FkYTg1YTJjYTg3NWNjMjciLCJwIjoiaiJ9

See updated [src/doc/figures/mission-states.png](https://github.com/jaiarobotics/jaiabot/blob/b2ab5f14655b1233d7e96a791b59a450b4702a02/src/doc/figures/mission-states.png) figure for details.

## Details 
- Moved all instances of IMURestart to new InMission::Pause::IMURestart state.
- Moved all instances of ReacquireGPS (except Dive::ReacquireGPS as the logic here is somewhat different) to new InMission::Pause::ReacquireGPS
- All Pause::* substates transition back to InMission using deep history so the mission will pick up from where it was paused.
- Renamed AcquiredGPSCommon to IvPSensorPauseCommon and updated this parent class to implement the reactions for transitioning to Pause::IMURestart and Pause::ReacquireGPS. 
- Add new Pause::Manual state and corresponding DCCL Command types (PAUSE/RESUME) to allow operator to manually pause and resume the mission for a given bot (**need JCC buttons for this**)
- Moved EvStop and EvAbort reactions from InMission::Underway to InMission so that they are still handled while in the new Pause state.

### Pause::ReacquireGPS
Added new SimulatorCommand to allow posting events to the simulator. Here, I added gps_dropout to allow injection of high hdop/pdop values for a set period of time.

Used the `goby zeromq publish` command to inject dropouts are various parts in the mission to ensure the simulator reacted correctly (transition to Pause::ReacquireGPS and then resume the mission from where it left off):
```
export GOBY_TOOL_LOAD_SHARED_LIBRARY=/home/toby/opensource/jaiabot/build/amd64/lib/libjaiabot_messages.so

goby zeromq publish jaiabot::simulator_command jaiabot.protobuf.SimulatorCommand 'gps_dropout { dropout_duration: 10 }' --interprocess 'platform: "bot1_fleet0"'
```

### Pause::IMURestart

Posted the IMUIssue message manually at various points in the simulated mission to ensure pause/resume worked correctly:

```
goby zeromq publish jaiabot::imu jaiabot.protobuf.IMUIssue 'solution: REBOOT_BNO085_IMU' --interprocess 'platform: "bot1_fleet0"'
```

### Pause::Manual

Used Goby Liaision to send 

```
bot_id: 1
time: 8536790027665155
type: PAUSE
```

and

```
bot_id: 1
time: 8536790052701920
type: RESUME
```

### CODE REVIEW

**Review by michael-jaia — CHANGES_REQUESTED** [review]
I think this feature will be a nice option for operators. @ferrom-jaia While playing with it in sim, I came across the following bug:
1. Create a run (assign a bot to it)
2. Play the run
3. After the run completes and the bot is station keeping, if you click the play all button, the bot will remain in a station keeping state (as opposed to restarting the run)

**Comment by ferrom-jaia** [comment]
> I think this feature will be a nice option for operators. @ferrom-jaia While playing with it in sim, I came across the following bug:
> 
> 1. Create a run (assign a bot to it)
> 2. Play the run
> 3. After the run completes and the bot is station keeping, if you click the play all button, the bot will remain in a station keeping state (as opposed to restarting the run)

This bug was already in 1.y. I have added a pull request to fix the issue. The described issue needs to have an unassigned run to reproduce error: https://github.com/jaiarobotics/jaiabot/pull/778

**Comment by tsaubergine** [comment]
@ferrom-jaia The Play button (which is now Pause) is useful for re-playing the mission once you reach the end. I would suggest switching it back to Play at the end of the mission (perhaps at IN_MISSION::UNDERWAY::RECOVERY).

**Comment by ferrom-jaia** [comment]
> @ferrom-jaia The Play button (which is now Pause) is useful for re-playing the mission once you reach the end. I would suggest switching it back to Play at the end of the mission (perhaps at IN_MISSION::UNDERWAY::RECOVERY).

Great idea! I will add that.

File: src/web/command_control/client/components/CommandControl.tsx [hunk 1]
```diff
@@ -2859,14 +2904,22 @@ export default class CommandControl extends React.Component {
 		if (botInfo.botIdsWifiDisconnected.length !==0) {
 			botInfo.downloadQueueMessage = `\nThe command cannot be sent to Bot${botInfo.botIdsWifiDisconnected.length > 1 ? 's': ''}: ` + botInfo.botIdsWifiDisconnected + " the Wi-Fi Link Quality is poor (Check Quick Look in Bot Details)"
 		}
+		if (botInfo.botIdsPauesNotAvaiable.length !==0) {
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: Let's add a space after the last equal sign

File: src/web/command_control/client/components/CommandControl.tsx [hunk 2]
```diff
@@ -1255,6 +1267,20 @@ export default class CommandControl extends React.Component {
 						doExecuteMission()
 					}
 
+					for (const botId of commDest.botIdsInPauseState) {
+						let c = {
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: Can we call the c variable command. I know in other parts of the code its named c but those should really be command. We can start the change by renaming this one command.

File: src/web/command_control/client/components/CommandControl.tsx [hunk 3]
```diff
@@ -1255,6 +1267,20 @@ export default class CommandControl extends React.Component {
...
+							bot_id: botId,
+							type: CommandType.RESUME
+						}
+				
+						console.log(c)
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: Similar thing here, we have console.log statements in other places but we probably shouldn't for production code. Do you agree?

File: src/web/command_control/client/components/CommandControl.tsx [hunk 4]
```diff
@@ -2765,12 +2799,14 @@ export default class CommandControl extends React.Component {
 	 * @param sendingActivate determines output and bot bots to send the acivate command to
 	 * @param sendingStop determines output and bot bots to send the stop command to
 	 * @param sendingDownload determines output and bot bots to send the download command to
+	 * @param sendingPause determines output and bot bots to send the pause command to
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: I think we only need bots here (I guess it's contagious as it comes from other the untouched code above haha)

File: src/web/command_control/client/components/CommandControl.tsx [hunk 5]
```diff
@@ -2823,6 +2862,12 @@ export default class CommandControl extends React.Component {
 				botInfo.botIdsDisconnected.push(bot?.bot_id)
 		    } else if (sendingDownload && !bot?.wifi_link_quality_percentage) {
 				botInfo.botIdsWifiDisconnected.push(bot?.bot_id)
+			} else if (sendingMission && pauseState.test(bot?.mission_state)) {
+				botInfo.botIdsInPauseState.push(bot?.bot_id)
+			} else if (sendingPause && (!stopAvailable.test(bot?.mission_state) ||
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: I am pretty confused by this else if block. Maybe it needs a comment or just a quick explanation in reply to this msg

File: src/web/command_control/client/components/CommandControl.tsx [hunk 6]
```diff
@@ -2859,14 +2904,22 @@ export default class CommandControl extends React.Component {
 		if (botInfo.botIdsWifiDisconnected.length !==0) {
 			botInfo.downloadQueueMessage = `\nThe command cannot be sent to Bot${botInfo.botIdsWifiDisconnected.length > 1 ? 's': ''}: ` + botInfo.botIdsWifiDisconnected + " the Wi-Fi Link Quality is poor (Check Quick Look in Bot Details)"
 		}
+		if (botInfo.botIdsPauesNotAvaiable.length !==0) {
+			botInfo.pauesNotAvaiableMessage  = `\nThe command cannot be sent to Bot${botInfo.botIdsPauesNotAvaiable.length > 1 ? 's': ''}: ` + botInfo.botIdsPauesNotAvaiable + ` because ${botInfo.botIdsPauesNotAvaiable.length > 1 ? 'they ': 'it '} in the incorrect state`
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: I think this is where we need to make it 'they are' and 'it is'

File: src/web/command_control/client/components/CommandControl.tsx [hunk 7]
```diff
@@ -2939,19 +2997,58 @@ export default class CommandControl extends React.Component {
...
+			const commDest = this.determineAllCommandBots(false, false, false, false, true)
+
+			if (commDest.botIds.length === 0) {
+				CustomAlert.alert(commDest.pauesNotAvaiableMessage + commDest.downloadQueueMessage + commDest.disconnectedMessage)
+			} else {
+				CustomAlert.confirm(`Click the OK button to pause Bot${commDest.botIds.length > 1 ? 's': ''}: ${commDest.botIds} ` +
+				commDest.pauesNotAvaiableMessage + commDest.downloadQueueMessage + commDest.disconnectedMessage, 'Pause All Bots', () => {
+
+					for (const botId of commDest.botIds) {
+						let c = {
```
Comment (michael-jaia) [inline src/web/command_control/client/components/CommandControl.tsx]: Wondering if we can do the same thing as above and name this variable command

File: src/web/command_control/client/components/Details.tsx [hunk 8]
```diff
@@ -114,6 +128,16 @@ const commands: {[key: string]: CommandInfo} = {
         humanReadableAvailable: "IN_MISSION__*, PRE_DEPLOYMENT__WAIT_FOR_MISSION_PLAN",
         humanReadableNotAvailable: ""
     },
+    resume: {
+        commandType: CommandType.RESUME,
+        description: 'Resume mission',
+        confirmationButtonText: 'Resume Mission',
+        statesAvailable: [
```
Comment (michael-jaia) [inline src/web/command_control/client/components/Details.tsx]: Should we add statesNotAvailable for consistency with the pause button, or do we even need that property for these btns?

**Review by michael-jaia — CHANGES_REQUESTED** [review]
@ferrom-jaia Leaving GitHub inline comments

**Review by ferrom-jaia — APPROVED** [review]
Looks great! Tested in sim with the procedures outlined in the pull request description (Super helpful!).

**Review by michael-jaia — DISMISSED** [review]
Tested web-side: The pause feature worked smoothly in sim. I focused on pausing individual bots, resuming individual bots, and the workings of the pause feature when a mission completes. Then I mixed in testing with the play all, pause-all and how it works when some bots are paused, moving, or stopped. The bots entered all of the expected states for the commands sent.

**Review by ferrom-jaia — CHANGES_REQUESTED** [review]
Need to do on the water testing before merging into 1.y

**Review by ferrom-jaia — DISMISSED** [review]
Removed UI code until we agree on a way forward. Backend works as expected.

**Review by ferrom-jaia — APPROVED** [review]

**Pull request merged**